### DISCUSSION TITLE
Help for implementation in a different setup.

### DISCUSSION THREAD

**Discussion by git2013vb** [Nyrest/FastGenericNew#20] [General]
Hi,
I have a specific setup where I'm using Activator.CreateInstance.
Because I have a memory leak I want to try your approach.

But I'm not an expert so I want to ask you help to understand how to do it.

Here my current code
I have an array of  class name like this one:

```
 public static readonly string[] authTables = new string[]
        {
            typeof(User).FullName,
            typeof(UserInfo).FullName
        };

```

and I give it to my method (in tables):

```
   public static void CreateTablesInDB(SqliteConnection con, string[] tables)
        {
            foreach (var table in tables)
            {
                Type tableType = Type.GetType(table + ", MyAssemblyName");//* assembly name                
                var instanceTable = Activator.CreateInstance(tableType);
                // here do some stuff with my instanceTable....
                string strCreateTable = ClassToDDL(instanceTable, "create");
            }
        }

```
User and UserInfo are my classes.

My question is: Can I use your approach? In which way?

Thank you.

**Comment by Nyrest** [comment]
Hi, thanks for the comment.

If `T` is unknown at compile time.
You can use `FastNew.GetCreateInstance<object>(type)`

Try:
```cs
 public static readonly Func<object>[] authTables = new Func<object>[]
        {
            FastNew.GetCreateInstance<object>(Type.GetType(typeof(User).FullName + ", MyAssemblyName")),
			FastNew.GetCreateInstance<object>(Type.GetType(typeof(UserInfo).FullName + ", MyAssemblyName"))
        };

 public static void CreateTablesInDB(SqliteConnection con, Func<object>[] tables)
        {
            foreach (var table in tables)
            {              
                var instanceTable = table();
                // here do some stuff with my instanceTable....
                string strCreateTable = ClassToDDL(instanceTable, "create");
            }
        }

```

**Comment by git2013vb** [comment]
Hi,
Thank you for your answer

I made the changes and I have memory leak when I call `var instanceTable = table();`

Any suggestions?

Thank you;

**Comment by Nyrest** [comment]
> Hi, Thank you for your answer
> 
> I made the changes and I have memory leak when I call `var instanceTable = table();`
> 
> Any suggestions?
> 
> Thank you;

Is the object you created implemented `IDisposable` or `IAsyncDisposable`?
If so you should call `Dispose()` or `DisposeAsync()` after your task is done.

```cs
public static void CreateTablesInDB(SqliteConnection con, Func<object>[] tables)
{
    foreach (var table in tables)
    {              
        var instanceTable = table();
        // here do some stuff with my instanceTable....
        string strCreateTable = ClassToDDL(instanceTable, "create");
        
        (instanceTable as IDisposable)?.Dispose();
    }
}
```

It can also be caused by the constructor code of the object.
You may need to decompile the DLL file that contains your object for more info.

And here are some tips about how to fix memory leaking problems:
https://docs.microsoft.com/en-us/dotnet/core/diagnostics/debug-memory-leak
https://michaelscodingspot.com/find-fix-and-avoid-memory-leaks-in-c-net-8-best-practices/

Hope this can help you out :)

**Comment by git2013vb** [comment]
Hi, thank you for your answer :)

It implement `IDisposable`

I added it and I added `GC.Collect();` after it.
Still have the memory leak. But I'm not sure if the tool I have keep the warning even if the memory will became free again. 
I have to investigate. 
Thank you for the links. I will read them. 

Sadly I'm working in Linux so I don't have VisualStudio installed. Instead I use VSCode.

I will came back to you when I will have more news.

Thank you again.
